src/norm/mongodb.nim

```diff
@@ -0,0 +1,727 @@
+
+const NORM_UNIVERSAL_TYPE_LIST* = @[
+  "float",
+  "string",
+  "Oid",
+  "bool",
+  "Time",
+  "int",
+  "Option[float]",
+  "Option[string]",
+  "Option[Oid]",
+  "Option[bool]",
+  "Option[Time]",
+  "Option[int]",
+  "seq[float]",
+  "seq[string]",
+  "seq[Oid]",
+  "seq[bool]",
+  "seq[Time]",
+  "seq[int]"
+]
+
+proc getCollectionName*(objRepr: ObjRepr): string =
+  ##[ Get the name of the DB table for the given object representation:
+  ``table`` pragma value if it exists or lowercased type name otherwise.
+  ]##
+
+  result = objRepr.signature.name.toLowerAscii()
+
+  for prag in objRepr.signature.pragmas:
+    if prag.name == "table" and prag.kind == pkKval:
+      return $prag.value
+
+
+proc getCollectionName*(T: typedesc): string =
+  ##[ Get the name of the DB table for the given type: ``table`` pragma value if it exists
+  or lowercased type name otherwise.
+  ]##
+
+  when T.hasCustomPragma(table): T.getCustomPragmaVal(table)
+  else: ($T).toLowerAscii()  
+
+
+var
+  allCollections*: seq[string] = @[]
+  normConnection*: string
+  normUser*: string
+  normPassword*: string
+  normDatabase*: string
+
+template genWithoutDb(newCollections: seq[string]): untyped {.dirty.} =
+  ## simply add to global allCollections
+
+  allCollections.insert(newCollections)
+
+template genWithDb(connection, user, password, database: string, useParams: int, newCollections: seq[string]): untyped {.dirty.} =
+  ## Generate ``withDb`` templates.
+
+  allCollections.insert(newCollections)
+  if useParams == -1:
+    normConnection = connection
+    normUser = user
+    normPassword = password
+    normDatabase = database
+
+  template withCustomDb*(
+    customConnection, customUser, customPassword, customDatabase: string,
+    body: untyped
+  ): untyped {.dirty.} =
+    ##[ A wrapper for actions that require custom DB connection, i.e. not the one defined in ``db``.
+    Defines CRUD procs to work with the DB.
+
+    'connection' should contain the URI pointing to the MongoDB server.
+
+    The 'user' and 'password' parameters are not used.
+
+    'database' should contain the database shard.
+
+    Aforementioned procs and procs defined in a ``db`` block can be used only
+    in  a ``withDb`` block.
+    ]##
+
+    block:
+      var dbConn = newMongoWithURI(customConnection)
+      let
+        dbConnResult = dbConn.connect()  ## TODO: Add more error handling
+
+      template dropTables() {.used.} =
+        ## Drops the collections for ALL objects.
+
+        for c in allCollections:
+          let dbCollection = dbConn[customDatabase][c]
+          let removeResult = dbCollection.drop()
+
+      template createTables(force = false) {.used.} =
+        ##[ For other database types, this function create new tables. However,
+        for MongoDB this has less meaning since collections are auto-created.
+        However, if `force` is set to true, it will dropTables first, deleting all
+        the documents.
+        ]##
+
+        if force:

```

Comment:
Maybe put a couple of debug log statements here, similar to how SQL queries are logged in other backends? Like, “Collections are created automatically, skipping.”

Reply:
Good idea. I will look into that with the failure scenario work soon.